hazelcast/src/main/java/com/hazelcast/config/IndexType.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.config;
+
+/**
+ * Type of the index.
+ */
+public enum IndexType {
+    /** Sorted index. Can be used with equality and range predicates. */
+    SORTED(0),
+
+    /** Hash index. Can be used with equality predicates. */
+    HASH(1);
+
+    private final int id;
+
+    IndexType(int id) {
+        this.id = id;
+    }
+
+    /**
+     * Gets the ID for the given {@link IndexType}.
+     *
+     * @return the ID
+     */
+    public int getId() {
+        return id;
+    }
+
+    /**
+     * Returns the IndexType as an enum.
+     *
+     * @return the IndexType as an enum
+     */
+    public static IndexType getById(final int id) {
+        for (IndexType type : values()) {
+            if (type.id == id) {
+                return type;
+            }
+        }
+        return null;
```

Comment:
Probably it makes sense to throw an exception here, WDYT?

Reply:
The current implementation is in line with other enums we have on config, except for `ConsistencyCheckStrategy`. I thought about throwing an exception also but then refused this idea because it looks more prone to hangs. 

The key thing is that normally this method should never return `null`. It is only possible if we messed up with compatibility. If we throw an exception from the `getById()` method, then this exception will be propagated in IO thread and we will not be able to deserialize the operation, what leads to a hang. On the other hand, if we allow `null` during serialization, then we allow for serialization to finish, and then we will have a kind of NPE inside operation thread, from which it is possible to send an error response back. 

Quick simulation in tests confirms that - an exception in IO thread leads to immediate hangs, while an NPE in operation thread allows for most tests to finish.

But again - this should not happen unless we have faulty compatibility in future releases.